From 4fb5bf9d1ee88ae112901ec4f3fef262830d2d97 Mon Sep 17 00:00:00 2001
From: Avinash Kumar <avikumar@habana.ai>
Date: Mon, 07 Jun 2021 13:54:55 +0300
Subject: [PATCH] [SW-31046] greco: scl low i2c slave in debug mode

Before power-up Soc BMC assert HOST_GPIO to LOW level.
Preboot fw monitors HOST_GPIO as follows:
LOW level- Greco in debug mode. Fw will wait till
HOST_GPIO is asserted to HIGH.
HIGH level- FW will read and write to modules and
monitor HOST_GPIO level.

When user done reading/writing from Soc, he must assert
HOST_GPIO to high level. For greco host_gpio is GPIO[44].

Change-Id: I98c3aa65a252f208b67b5e80b9b5d97359f44410
Signed-off-by: Avinash Kumar <avikumar@habana.ai>
---

diff --git a/arch/greco/common.c b/arch/greco/common.c
index 732aa0c..6348b1c 100644
--- a/arch/greco/common.c
+++ b/arch/greco/common.c
@@ -12,6 +12,7 @@
 #include <psoc_global_conf_regs.h>
 #include <psoc_global_conf_masks.h>
 #include <psoc_efuse_regs.h>
+#include <psoc_gpio2_regs.h>
 #include <pcie_aux_regs.h>
 #include <pcie_core_regs.h>
 #include <pcie_wrap_regs.h>
@@ -51,6 +52,25 @@
 #define FUSE_READ_RANGE_ALL	0x3F80
 #define PCIE_LTR_DISABLE		0
 #define PCIE_CAP_HDR_NXT_OFFSET		0x1f8
+#define HOST_GPIO_I2C_DEBUG	44
+
+/* HOST_GPIO is set through mmPSOC_GPIO2, each mmPSOC_GPIO serve 16 GPIOs.*/
+bool preboot_host_gpio_asserted(void)
+{
+	u32 reg, mask;
+
+	mask = BIT(HOST_GPIO_I2C_DEBUG % 16);
+
+	/* configure as input */
+	regmap_writel(mask, mmPSOC_GPIO2_OUTENCLR);
+
+	/* clear ALT function for gpio */
+	regmap_writel(mask, mmPSOC_GPIO2_ALTFUNCCLR);
+
+	reg  = regmap_readl(mmPSOC_GPIO2_DATA);
+
+	return reg & mask;
+}
 
 u32 get_efuse_read_reg(void)
 {
diff --git a/arch/greco_arc_gnu/common.c b/arch/greco_arc_gnu/common.c
index 5d73a22..ed3d01a 100644
--- a/arch/greco_arc_gnu/common.c
+++ b/arch/greco_arc_gnu/common.c
@@ -10,6 +10,7 @@
 #include <pcie_phy_regs.h>
 #include <psoc_global_conf_regs.h>
 #include <psoc_gpio1_regs.h>
+#include <psoc_gpio2_regs.h>
 #include <psoc_qspi_regs.h>
 #include <psoc_efuse_regs.h>
 #include <psoc_arc0_cfg_regs.h>
@@ -45,6 +46,25 @@
 
 #define mmPCIE_WRAP_VUART_TX_DATA       0xC15F60
 #define mmPCIE_WRAP_VUART_TX_STATUS     0xC15F64
+#define HOST_GPIO_I2C_DEBUG	44
+
+/* HOST_GPIO is set through mmPSOC_GPIO2, each mmPSOC_GPIO serve 16 GPIOs.*/
+bool preboot_host_gpio_asserted(void)
+{
+	u32 reg, mask;
+
+	mask = BIT(HOST_GPIO_I2C_DEBUG % 16);
+
+	/* configure as input */
+	regmap_writel(mask, mmPSOC_GPIO2_OUTENCLR);
+
+	/* clear ALT function for gpio */
+	regmap_writel(mask, mmPSOC_GPIO2_ALTFUNCCLR);
+
+	reg  = regmap_readl(mmPSOC_GPIO2_DATA);
+
+	return reg & mask;
+}
 
 u32 get_efuse_read_reg(void)
 {
diff --git a/common/cmd_module.c b/common/cmd_module.c
index 8e5e2a5..484f22c 100644
--- a/common/cmd_module.c
+++ b/common/cmd_module.c
@@ -10,6 +10,26 @@
 #include "uart.h"
 #include "cmd_module.h"
 
+static void i2c_debug_mode_handling(void)
+{
+	bool debug_mode;
+
+	debug_mode = hl_is_debug_mode(HL_DEBUG_HW_I2C);
+
+	/* i2c debug mode is true and gpio is deasserted */
+	if ((debug_mode == true) &&
+		 (preboot_host_gpio_asserted() == false)) {
+
+		/*
+		 * Wait till gpio is asserted. When gpio is de-asserted
+		 * its in debug mode. when gpio is asserted fw will
+		 * read write to modules.
+		 */
+		while (!preboot_host_gpio_asserted())
+			;
+	}
+}
+
 void module_handle_cmds(struct cmd_module_ops cmd_ops[], int size)
 {
 	int i = 0, j = 0;
@@ -19,8 +39,12 @@
 		cmd_ops[j].setup();
 
 	preboot_puts("Waiting for commands...\n");
+
 	i = 0;
 	while (1) {
+
+		i2c_debug_mode_handling();
+
 		res = cmd_ops[i].check();
 		switch (res) {
 		case NOT_FOUND:
diff --git a/include/preboot.h b/include/preboot.h
index 560d6f2..d385893 100644
--- a/include/preboot.h
+++ b/include/preboot.h
@@ -145,6 +145,7 @@
 void preboot_pcie_enable(void);
 void preboot_pcie_exec_hw_wa(void);
 void preboot_nic_en(u32 nic_en_);
+bool preboot_host_gpio_asserted(void);
 
 void ppboot_cp_preboot(u64 loadaddr, u32 offset, u32 size, char *name);
 void image_start(u64 loadaddr);
diff --git a/preboot/preboot.c b/preboot/preboot.c
index 201fe2e..fb1b7d6 100644
--- a/preboot/preboot.c
+++ b/preboot/preboot.c
@@ -248,6 +248,7 @@
 __weak void preboot_led_on(void) {}
 __weak void preboot_sp_sram_init(void) {}
 __weak void hl_ram_init(void) {}
+__weak bool preboot_host_gpio_asserted(void) { return true; }
 
 static bool preboot_is_overlap_address_range(u64 addr, int len)
 {
